src/main/java/com/sparrowwallet/drongo/wallet/bip93/Codex32.java

```diff
@@ -0,0 +1,225 @@
+package com.sparrowwallet.drongo.wallet.bip93;
+
+import com.sparrowwallet.drongo.protocol.Bech32;
+import com.sparrowwallet.drongo.protocol.ProtocolException;
+import com.sparrowwallet.drongo.wallet.MnemonicException;
+
+import java.math.BigInteger;
+import java.util.Arrays;
+import java.util.Locale;
+
+public class Codex32 {
+
+    private static final String HRP = "ms";
+    private static final char SPECIAL_SHARE_INDEX = 's';
+    private static final int CODEX32_ID_LEN = 4;
+
+    public static class Codex32Data {
+        public final byte[] rawData;
+        public final ChecksumType checksumType;
+
+        public Codex32Data(byte[] rawData, ChecksumType checksumType) {
+            this.rawData = rawData;
+            this.checksumType = checksumType;
+        }
+
+        public byte getThreshold() {
+            return rawData[0];
+        }
+
+        public int getThresholdAsInt() {
+            return Bech32.CHARSET.charAt(getThreshold()) - '0';
+        }
+
+        public byte[] getIdentifier() {
+            return Arrays.copyOfRange(rawData, 1, 5);
+        }
+
+        public String identifierAsString() {
+            byte[] id = getIdentifier();
+
+            StringBuilder sb = new StringBuilder(CODEX32_ID_LEN);
+            for(byte b : id) {
+                sb.append(Bech32.CHARSET.charAt(b));
+            }
+            return sb.toString();
+        }
+
+        public byte getShareIndex() {
+            return rawData[5];
+        }
+
+        public boolean isUnsharedSecret() {
+            return getShareIndex() == Bech32.CHARSET_REV[SPECIAL_SHARE_INDEX];
+        }
+
+        public byte[] getPayload() {
+            return Arrays.copyOfRange(rawData, 6, rawData.length);
+        }
+
+        public byte[] payloadToBip32Secret() throws MnemonicException {
+            if(!isUnsharedSecret()) {
+                throw new MnemonicException("Trying to get secret from non-secret share");
+            }
+            byte[] payload = getPayload();
+            return Bech32.convertBits(payload, 0, payload.length, 5, 8, false, false);
+        }
+    }
+
+    public static String encode(Codex32Data data) {
```

Comment:
Some validation would be good to perform before encoding, mainly to check the length of `rawData` and that the matching checksum type has been specified.